### DISCUSSION TITLE
How to Handle Longer Data Set

### DISCUSSION THREAD

**Discussion by abbasmashaddy72** [livewire/livewire#5240] [1. Help]
I have created a depended on dropdown which is working fine without any issues

Ex: 1) Locality Depends on City, 2) City Depends on State & 3) States Independent

But My Client Wants to select direct from the locality dropdown when I am trying to do that, I am getting:

Error: This page isn’t working

I know What the issue is, Locality has almost 90K records which is unable to handle it
City: 7K
State: 36

I also used `wire:init` so that I can load the data after page is loaded but unfortunately no success in it

Please Help me Out how can I handle that much data

I had an Idea but Was not able to implement it

Idea: Search live data using where Condition Like Query But didnot work

LiveWire Component
`<?php

namespace App\Http\Livewire\Components;

use App\Models\City;
use App\Models\Locality;
use App\Services\Helper;
use Livewire\Component;

class StateCityLocality extends Component
{
    public $states = [];
    public $cities = [];
    public $localities = [];

    public $selectedState = null;
    public $selectedCity = null;
    public $selectedLocality = null;

    public function mount($selectedLocality = null)
    {
        $this->states = Helper::getKeyValuesWithMap('State', 'name', 'id');

        if (!is_null($selectedLocality)) {
            $locality = Locality::find($selectedLocality);
            $this->localities = Helper::getKeyValuesWithMap('Locality', 'name', 'id', 'city_id', $locality->city_id);
            $city = City::find($locality->city_id);
            $this->cities = Helper::getKeyValuesWithMap('City', 'name', 'id', 'state_id', $city->state_id);
            $this->selectedCity = $locality->city_id;
            $this->selectedState = $city->state_id;
        }
    }

    public function render()
    {
        return view('livewire.components.state-city-locality');
    }

    public function updatedSelectedState($state)
    {
        $this->cities = Helper::getKeyValuesWithMap('City', 'name', 'id', 'state_id', $state);
        $this->selectedCity = null;
    }

    public function updatedSelectedCity($city)
    {
        $this->localities = Helper::getKeyValuesWithMap('Locality', 'name', 'id', 'city_id', $city);
    }

    public function updatedSelectedLocality()
    {
        $this->emit('locality_changed', ['locality_id' => $this->selectedLocality]);
    }
}`

LiveView Component View
`<div>

    <div class="grid-cols-3 gap-2 sm:grid">

        <x-simple-select name="selectedLocality" id="selectedLocality" label="Locality" wire:model="selectedLocality"
            :options="$localities" value-field='id' text-field='name' placeholder="Select Locality"
            search-input-placeholder="Search Locality" :searchable="true" />

        <x-simple-select name="selectedCity" id="selectedCity" label="City" wire:model="selectedCity" :options="$cities"
            value-field='id' text-field='name' placeholder="Select City" search-input-placeholder="Search City"
            :searchable="true" />

        <x-simple-select name="selectedState" id="selectedState" label="State" wire:model="selectedState"
            :options="$states" value-field='id' text-field='name' placeholder="Select State"
            search-input-placeholder="Search State" :searchable="true" />
    </div>

</div>`

My Simple Select Component Component
`props([
    'label' => '',
])
<div x-cloak class="relative mt-4" x-data="SimpleSelect({
    dataSource: {{ is_array($options) ? json_encode($options) : json_encode([]) }},
    @if ($attributes->whereStartsWith('wire:model')->first()) selected: @entangle($attributes->wire('model')),
        @else
            selected: '', @endif
    valueField: '{{ $valueField }}',
    textField: '{{ $textField }}',
    value: @if (is_array($value)) {{ json_encode($value) }} @else '{{ $value }}' @endif,
    name: '{{ $name }}',
    id: '{{ $id }}',
    placeholder: '{{ $placeholder }}',
    searchInputPlaceholder: '{{ $searchInputPlaceholder }}',
    noOptions: '{{ $noOptions }}',
    noResult: '{{ $noResult }}',
    multiple: {{ isset($attributes['multiple']) ? 'true' : 'false' }},
    maxSelection: '{{ $maxSelection }}',
    required: {{ isset($attributes['required']) ? 'true' : 'false' }},
    disabled: {{ isset($attributes['disabled']) ? 'true' : 'false' }},
    searchable: {{ $searchable ? 'true' : 'false' }},
    clearable: {{ $clearable ? 'true' : 'false' }},
    onSelect: '{{ $attributes['on-select'] ?? 'select' }}'
})" x-init="init();" x-on:click.outside="closeSelect()"
    x-on:keydown.escape="closeSelect()" :wire:key="`${id}${generateID()}`">
    <x-form-label label="{{ $label }}" />
    <div x-ref="simpleSelectButton" x-on:click="toggleSelect()" x-on:keyup.enter="toggleSelect()" tabindex="0"
        x-bind:class="{
            'rounded-md': !open,
            'rounded-t-md': open,
            'bg-gray-200 cursor-default': disabled
        }"
        {{ $attributes->class('block w-full border border-state-200 rounded-md shadow-sm focus:ring-0 focus:ring-gray-400 focus:border-gray-400 sm:text-sm sm:leading-5 mt-1 dark:bg-darkmode-800')->only('class') }}>
        <div x-show="!selected || selected.length === 0" class="flex flex-wrap">
            <div class="text-gray-800 rounded-sm w-full truncate px-2 py-1 my-0.5 flex flex-row items-center">
                <div class="w-full px-2 truncate dark:text-slate-300" x-text="placeholder">&nbsp;</div>
                <div x-show="!disabled" x-bind:class="{ 'cursor-pointer': !disabled }" class="h-6"
                    x-on:click.prevent.stop="toggleSelect()">
                    @include('simple-select::components.caret-icons')
                </div>
            </div>
        </div>
        @isset($attributes['multiple'])
            <div x-show="selected && typeof selected === 'object' && selected.length > 0" class="flex flex-wrap space-x-1">
                <template x-for="(value, index) in selected" :key="index">
                    <div
                        class="text-gray-800 dark:text-gray-400 rounded-full truncate bg-gray-300 dark:bg-gray-800 px-2 py-0.5 my-0.5 flex flex-row items-center">
                        {{-- Invisible inputs for standard form submission values --}}
                        <input type="text" :name="`${name}[]`" x-model="value" style="display: none;" />
                        <div class="px-2 truncate">
                            @isset($customSelected)
                                <span x-data="{ option: getOptionFromSelectedValue(value) }">
                                    <template
                                        x-if="(typeof option === 'string' && option.length > 0) || (typeof option === 'object' && Object.keys(option).length > 0)">
                                        <span>{{ $customSelected }}</span>
                                    </template>
                                </span>
                            @else
                                <span x-text="getTextFromSelectedValue(value)"></span>
                            @endisset
                        </div>
                        <div x-show="!disabled" x-bind:class="{ 'cursor-pointer': !disabled }"
                            x-on:click.prevent.stop="deselectOption(index)" x-on:keyup.enter="deselectOption(index)"
                            class="w-6" tabindex="0">
                            @include('simple-select::components.deselect-icon')
                        </div>
                    </div>
                </template>
            </div>
        @else
            <div x-show="selected" class="flex flex-wrap">
                <div
                    class="text-gray-800 dark:text-slate-300 rounded-sm w-full truncate px-2 py-1 my-0.5 flex flex-row items-center">
                    {{-- Invisible input for standard form submission of values --}}
                    <input type="text" :name="name" x-model="selected" :required="required"
                        style="display: none;" />
                    <div class="w-full px-2 truncate">
                        @isset($customSelected)
                            <span x-data="{ option: getOptionFromSelectedValue(selected) }">
                                <template
                                    x-if="option !== null && ((typeof option === 'string' && option.length > 0) || (typeof option === 'object' && Object.keys(option).length > 0))">
                                    <span>{{ $customSelected }}</span>
                                </template>
                            </span>
                        @else
                            <span x-text="getTextFromSelectedValue(selected)"></span>
                        @endisset
                    </div>
                    <div x-show="!disabled && clearable" x-bind:class="{ 'cursor-pointer': !disabled }"
                        x-on:click.prevent.stop="deselectOption()" x-on:keyup.enter="deselectOption()" class="h-6"
                        tabindex="0">
                        @include('simple-select::components.deselect-icon')
                    </div>
                    <div x-show="!disabled && !clearable" x-bind:class="{ 'cursor-pointer': !disabled }" class="h-6"
                        tabindex="0">
                        @include('simple-select::components.caret-icons')
                    </div>
                </div>
            </div>
        @endisset
    </div>
    <div x-ref="simpleSelectOptionsContainer" x-bind:style="open ? 'height: ' + popperHeight : ''"
        class="absolute z-10 w-full">
        <div x-show="open">
            <input type="search" x-show="searchable" x-ref="simpleSelectOptionsSearch" x-model="search"
                x-on:click.prevent.stop="open=true" :placeholder="searchInputPlaceholder"
                class="block w-full p-2 bg-gray-100 border border-gray-300 shadow-md dark:bg-gray-800 dark:border-gray-700 dark:text-gray-200 focus:border-gray-200 focus:ring-0 sm:text-sm sm:leading-5" />
            <ul x-ref="simpleSelectOptionsList"
                class="absolute z-10 w-full py-1 overflow-auto text-base bg-white shadow-lg dark:bg-gray-600 rounded-b-md max-h-60 ring-1 ring-black ring-opacity-5 focus:outline-none sm:text-sm"
                tabindex="-1" role="listbox">
                <div x-show="Object.values(options).length == 0 && search.toString().trim() == ''" x-text="noOptions"
                    class="px-2 py-2 dark:text-gray-400"></div>
                <div x-show="Object.values(options).length == 0 && search.toString().trim() != ''" x-text="noResult"
                    class="px-2 py-2 dark:text-gray-400"></div>
                <template x-for="(option, index) in Object.values(options)" :key="index">
                    <li :tabindex="index" class="relative py-2 pl-3 select-none pr-9 whitespace-nowrap"
                        @isset($attributes['multiple'])
                            x-bind:class="{
                                'bg-gray-300 dark:bg-gray-700 text-black dark:text-gray-300 hover:none': selected && selected.includes(getOptionValue(option, index)),
                                'text-gray-900 dark:text-gray-400 cursor-defaul hover:bg-gray-200 dark:hover:bg-gray-800 hover:cursor-pointer focus:bg-gray-200': !(selected && selected.includes(getOptionValue(option, index))),
                            }"
                        @else
                            x-bind:class="{
                                'bg-gray-300 dark:bg-gray-700 text-black dark:text-gray-300 hover:none': selected == getOptionValue(option, index),
                                'text-gray-900 dark:text-gray-400 cursor-defaul hover:bg-gray-200 dark:hover:bg-gray-800 hover:cursor-pointer focus:bg-gray-200': !(selected == getOptionValue(option, index)),
                            }"
                        @endisset
                        x-on:click="selectOption(getOptionValue(option, index))"
                        x-on:keyup.enter="selectOption(getOptionValue(option, index))">
                        @isset($customOption)
                            {{ $customOption }}
                        @else
                            <span x-text="getOptionText(option, index)"></span>
                        @endisset
                    </li>
                </template>
            </ul>
        </div>
    </div>
</div>

<script>
    function SimpleSelect(config) {
      return {
          open: false,
          search: '',
          dataSource: config.dataSource,
          options: {},
          valueField: config.valueField,
          textField: config.textField,
          value: config.value,
          placeholder: config.placeholder,
          selected: config.selected,
          searchable: config.searchable,
          clearable: config.clearable,
          required: config.required,
          disabled: config.disabled,
          multiple: config.multiple,
          maxSelection: config.maxSelection,
          name: config.name,
          id: config.id,
          searchInputPlaceholder: config.searchInputPlaceholder,
          noOptions: config.noOptions,
          noResult: config.noResult,
          onSelect: config.onSelect,
          isLoading: false,
          popperInstance: null,
          popperHeight: '0px',
    
          init: function() {
              if (this.value && !this.selected) {
                  this.selected = this.value;
              }
              
              if (!this.selected) {
                if (this.multiple) {
                    this.selected = [];
                } else {
                    this.selected = null;
                }
              }
              this.resetOptions();
    
              this.$watch('search', ((search) => {
                  this.resetOptions();
                  this.options = Object.values(this.dataSource)
                      .filter((value, index) => this.getOptionText(value, index).toString().toLowerCase().includes(search.toLowerCase().trim()));
                  
                  setTimeout(() => {
                      this.popper();
                      this.scrollToOption();
                  }, 100);
              }));
          },
    
          generateID: () => {
              return '_' + Math.random().toString(36).substr(2, 9)
          },
          
          resetOptions: function(dataSource = null) {                
              if (!dataSource) {
                  dataSource = this.dataSource;
              }
              this.options = Object.values(dataSource);       
          },
    
          toggleSelect: function() {
              if (!this.disabled) {
                  if (this.open) {
                      this.closeSelect();
                  } else {
                      this.open = true;
                                              
                      setTimeout(() => {
                          this.popper();
                          this.scrollToOption();      
                          this.checkMaxSelectionReached();  
                      }, 700);          
                  }                      
              }
          },
    
          closeSelect: function() {
              this.open = false;
              this.search = '';
          },
    
          selectOption: function(value) {
              if(!this.disabled) {
                  // If multiple push to the array, if not, keep that value and close menu
                  if (this.multiple) {
                      // If it's not already in there
                      if (!this.selected.includes(value)) {
                          if (this.maxSelection == 0 || (this.maxSelection > 0 && this.selected.length < this.maxSelection)) {
                              this.selected.push(value)                                
                          }
                          let reached = this.checkMaxSelectionReached();
                          if (reached) {
                              this.closeSelect();
                          }
                      }
                  } else {
                      this.selected = value;
                      this.closeSelect();
                  }
                  if (this.onSelect) {
                      this.$dispatch(`${this.onSelect}`, { 
                          id: this.id,
                          name: this.name,
                          value: this.selected
                      });
                  }
              }
          },
          
          deselectOption: function(index = null) {
              if (this.multiple) {
                  this.selected.splice(index, 1)
              } else {
                  this.selected = '';
              }
          },
    
          checkMaxSelectionReached: function() {
              if (this.multiple && this.$refs.simpleSelectOptionsList) {
                  if (this.maxSelection > 0 && this.selected.length >= this.maxSelection) {
                      this.$refs.simpleSelectOptionsList.querySelectorAll(':scope > li').forEach((el) => {
                          el.style.pointerEvents = 'none';
                          el.style.opacity = '0.4';
                      });
                      return true;
                  } else {
                      this.$refs.simpleSelectOptionsList.querySelectorAll(':scope > li').forEach((el) => {
                          el.style.pointerEvents = 'auto';
                          el.style.opacity = '1';
                      });
                  }
              }
              return false;
          },
    
          getOptionValue: function(option, index) {
              return typeof option === 'object' && this.options[index] ? this.options[index][this.valueField] : option;
          },
    
          getOptionText: function(option, index) {
              return typeof option === 'object' && this.options[index] ? this.options[index][this.textField] : option;
          },
    
          getIndexFromSelectedValue: function(value) {    
              let valueField = this.valueField;
              return Object.values(this.dataSource).findIndex(function(x) {
                  if (typeof x === 'object') {
                      return x[valueField] == value;
                  } else {
                      return x == value;
                  }
              });       
          },
    
          getTextFromSelectedValue: function(value) {
              let index = this.getIndexFromSelectedValue(value);
              let valueField = this.valueField;
              let foundValue = Object.values(this.dataSource).find(function(x) {
                  if (typeof x === 'object') {
                      return x[valueField] == value;
                  } else {
                      return x == value;
                  }
              });
              return typeof foundValue === 'object' && this.dataSource[index] ? this.dataSource[index][this.textField] : foundValue;
          },
    
          getOptionFromSelectedValue: function(value) {
              let index = this.getIndexFromSelectedValue(value);
              let valueField = this.valueField;
              let foundValue = Object.values(this.dataSource).find(function(x) {
                  if (typeof x === 'object') {
                      return x[valueField] == value;
                  } else {
                      return x == value;
                  }
              });
              return foundValue ?? value;
          },
    
          popper: function() {
              // update popper position
              if (this.$refs.simpleSelectOptionsList && this.$refs.simpleSelectOptionsList.offsetHeight) {
                  this.popperHeight = (this.$refs.simpleSelectOptionsList.offsetHeight + 0) + 'px';
              }
              
              let createPopper = window.Popper ? window.Popper.createPopper : null;
              createPopper = !createPopper && window.createPopper ? window.createPopper : null;
    
              if (typeof createPopper !== 'function') {
                  throw new TypeError('Laravel Simple Select: requires Popper (https://popper.js.org)');
              }
                          
              if (createPopper && this.$refs.simpleSelectButton && this.$refs.simpleSelectOptionsContainer) {
                  this.popperInstance = createPopper(this.$refs.simpleSelectButton, this.$refs.simpleSelectOptionsContainer, {
                      // placement: "auto",
                      placement: "bottom-start",
                      modifiers: [
                          {
                              name: 'offset',
                              options: {
                                  offset: [0, 0],
                              },
                          },
                          {
                              name: "preventOverflow",
                              options: { 
                                  boundary: "clippingParents" 
                              },
                          },
                          { 
                              name: "flip", 
                              options: { 
                                  padding: 20,
                                  allowedAutoPlacements: ['top', 'bottom'],
                              }
                          }
                      ]
                  });
              }
          },
    
          scrollToOption: function () {                
              try {
                  if (this.selected && this.$refs.simpleSelectOptionsList) {
                      let focusIndex = 0;
                      if (this.multiple) {
                          let lastSelected = this.selected.length > 0 ? this.selected[this.selected.length - 1] : '';
                          focusIndex = lastSelected ? this.getIndexFromSelectedValue(lastSelected) : 0;
                      } else {
                          focusIndex = this.getIndexFromSelectedValue(this.selected);
                      }                        
                      // let nonListItem = 3;
                      // let totalListItem = this.$refs.simpleSelectOptionsList.children.length > nonListItem ? this.$refs.simpleSelectOptionsList.children.length - nonListItem : 0;
                      let optionsList = this.$refs.simpleSelectOptionsList.querySelectorAll(':scope > li');
                      let totalOptionsList = optionsList.length;
                      if (totalOptionsList > 0) {
                          let offsetTop = optionsList[focusIndex].offsetTop;
                          this.$refs.simpleSelectOptionsList.scrollTop = offsetTop || 0;
                          // optionsList[focusIndex].focus();                            
                      } else {
                          this.$refs.simpleSelectOptionsList.scrollTop = 0;                            
                      }                        
                  } else {
                      this.$refs.simpleSelectOptionsList.scrollTop = 0;
                  }
                  this.$refs.simpleSelectOptionsSearch.focus();
              } catch (e) {}                    
              
          }
      }
    }
    window.SimpleSelect = SimpleSelect;
    </script>`

Please Help me Out

Thanks in Advance

**Comment by abbasmashaddy72** [comment]
Please Help me out

@andreascreten @wuwx @inxilpro @Treggats

**Reply by Treggats** [reply]
Hello,

You need to format your question better, this isn't very readable. I can only respond to your actual question.

Regarding your question, it seems like you already know what your issue is. Too many records for the dropdown.
If you need that many records then a dropdown isn't a really good solution.

You'd better have a search input for which you do a like query on every three (for example) characters. Just like you already thought of.

Hope this helps

**Reply by abbasmashaddy72** [reply]
Can give an Idea How can I implement it.

**Reply by Treggats** [reply]
There are plenty resources available that you can use to build this.

**Reply by abbasmashaddy72** [reply]
Thanks I will check it out
